datadog_checks_base/datadog_checks/base/checks/base.py

```diff
@@ -132,48 +135,55 @@ class except the :py:meth:`check` method but sometimes it might be useful for a
                 configuration file (a list is used to keep backward compatibility with
                 older versions of the Agent).
         """
-        self.metrics = defaultdict(list)
+        self.metrics = defaultdict(list)  # type: typing.DefaultDict[str, list]
         self.check_id = ''
-        self.instances = kwargs.get('instances', [])
-        self.name = kwargs.get('name', '')
-        self.init_config = kwargs.get('init_config', {})
-        self.agentConfig = kwargs.get('agentConfig', {})
-        self.warnings = []
-        self.metric_limiter = None
+        self.warnings = []  # type: typing.List[str]
+        self.metric_limiter = None  # typing.Optional[Limiter]
+
+        instances = kwargs.get('instances', [])
+        name = kwargs.get('name', '')
+        init_config = kwargs.get('init_config', {})
+        agentConfig = kwargs.get('agentConfig', {})
 
         if len(args) > 0:
-            self.name = args[0]
+            name = args[0]
         if len(args) > 1:
-            self.init_config = args[1]
+            init_config = args[1]
         if len(args) > 2:
             # agent pass instances as tuple but in test we are usually using list, so we are testing for both
             if len(args) > 3 or not isinstance(args[2], (list, tuple)) or 'instances' in kwargs:
                 # old-style init: the 3rd argument is `agentConfig`
-                self.agentConfig = args[2]
+                agentConfig = args[2]
                 if len(args) > 3:
-                    self.instances = args[3]
+                    instances = args[3]
             else:
                 # new-style init: the 3rd argument is `instances`
-                self.instances = args[2]
+                instances = args[2]
+
+        # The assignments above are a bit too dynamic for mypy, so let's give it some help.
+        self.instances = typing.cast(typing.Sequence[typing.Dict[str, typing.Any]], instances)
+        self.name = typing.cast(str, name)
+        self.init_config = typing.cast(typing.Dict[str, typing.Any], init_config)
+        self.agentConfig = typing.cast(typing.Dict[str, typing.Any], agentConfig)
```

Comment:
A bit -1 to start to do that. I'm fine with typing hints with comments, but it goes too far into making the code change.